### DISCUSSION TITLE
How to fake HttpClient requests

### DISCUSSION THREAD

**Discussion by vpandiri** [jordimontana82/fake-xrm-easy#544] [Q&A] [answered]
Hello everyone,
  We have some plugins and custom workflow activities which calls external APIs using HttpClient. Is there a way to fake these external API calls when using FakeXrmEasy? 

Thanks
Vikranth P

**Comment by jordimontana82** [comment, marked as answer]
Hi Vikranth,

What I would normally do is:

1 - Add a public property of an interface of that service to the plugin/cwa like: 
    
    public IHttpClient myService;

2 - In the unit test, setup a specific plugin/cwa instance passing in a mock of that property (i.e. with FakeItEasy)

3 - Call the ExecutePluginWith / ExecuteCodeActivity method that allows you to pass in that plugin/cwa instance

4 - In the plugin / cwa 's .Execute method, add an if statement like the below so it would only create a real instance when not run from the unit test project (i.e. when  runs against the real CRM/CDS/ instance)

    if(myService == null) {
           //create an instance with real IP address and so on....
    }

5 - The rest of the plugin/cwa code to reference the interface (property) only.

Note: adding plugin / cwa cause the approach is the same for both plugins and custom workflow activities...

**Reply by vpandiri** [reply]
Thanks @jordimontana82 for quick reply and suggestion. I will try this approach and let you know if it works. 
Also, instead of updating plugin code is there a way to inject fake handler for http requests?

**Reply by jordimontana82** [reply]
Since plugins only have a parameterless constructor , then no, I'm not aware of any other way apart from injecting properties.

**Reply by vpandiri** [reply]
Thanks @jordimontana82 . Followed your proposed approach and it works for us. Appreciate your quick help. Thank you.
